#VRML_SIM R2023b utf8
PROTO robotModel [
  field SFVec3f    translation    2.71944 -0.409782 0.143814
  field SFRotation rotation       -0.0057741504742595436 -0.023588301937423933 0.9997050821107241 3.07635
  field SFString   controller     "bug_2_controller"
]
{
  Robot {
    translation IS translation
    rotation IS rotation
    controller IS controller
    supervisor TRUE
    children [
      Compass {
        translation 0 0 0.05
        rotation 0 0 1 1.5708
      }
      Lidar {
        translation 0 0 0.06
        rotation 0 0 1 1.5708
        name "lidar_front"
        rotationStep 0.261789
        fieldOfView 1
        numberOfLayers 1
        maxRange 0.5
      }
      Lidar {
        translation 0 0 0.06
        rotation 0 0 1 3.14159
        name "lidar_left"
        rotationStep 0.2618
        horizontalResolution 256
        fieldOfView 0.7854
        numberOfLayers 1
        maxRange 0.6
      }
      Lidar {
        translation 0 0 0.05
        children [
          Shape {
            appearance PBRAppearance {
              roughness 1
              metalness 0
            }
            geometry Box {
              size 0.02 0.02 0.02
            }
          }
        ]
        name "lidar_right"
        horizontalResolution 256
        fieldOfView 0.7854
        numberOfLayers 1
        maxRange 0.6
      }
      GPS {
        translation 0 0 0.05
      }
      DistanceSensor {
        translation -1.10196e-07 0.1 0.03
        rotation 0 0 1 1.5708
        children [
          Shape {
            appearance PBRAppearance {
              roughness 1
              metalness 0
            }
            geometry Box {
              size 0.01 0.01 0.01
            }
          }
        ]
        name "sonar_front"
        lookupTable [
          0 0 0
          0.43 5000 0
        ]
        type "sonar"
      }
      DistanceSensor {
        translation -0.0849986 -0.0500001 0.04
        rotation 0 0 1 -2.617985307179586
        children [
          Shape {
            appearance PBRAppearance {
              roughness 1
              metalness 0
            }
            geometry Box {
              size 0.01 0.01 0.01
            }
          }
        ]
        name "sonar_back_left"
        lookupTable [
          0 0 0
          0.5 5000 0
        ]
      }
      DistanceSensor {
        translation -0.1 0 0.04
        rotation 0 0 1 -3.1415853071795863
        children [
          Shape {
            appearance PBRAppearance {
              roughness 1
              metalness 0
            }
            geometry Box {
              size 0.01 0.01 0.01
            }
          }
        ]
        name "sonar_left"
        lookupTable [
          0 0 0
          0.5 5000 0
        ]
        type "sonar"
      }
      DistanceSensor {
        translation 0.085 -0.05 0.04
        rotation 0 0 -1 0.523596
        children [
          Shape {
            appearance PBRAppearance {
              roughness 1
              metalness 0
            }
            geometry Box {
              size 0.01 0.01 0.01
            }
          }
        ]
        name "sonar_back_right"
        lookupTable [
          0 0 0
          0.5 5000 0
        ]
        type "sonar"
      }
      DistanceSensor {
        translation 0.1 0 0.04
        rotation 0 0 1 2.77402e-06
        children [
          Shape {
            appearance PBRAppearance {
              roughness 1
              metalness 0
            }
            geometry Box {
              size 0.01 0.01 0.01
            }
          }
        ]
        name "sonar_right"
        lookupTable [
          0 0 0
          0.5 5000 0
        ]
        type "sonar"
      }
      # --- Rear wheel
      HingeJoint {
        jointParameters HingeJointParameters {
          position 78.21503859814996
          axis -1 0 0
          anchor 0 -0.04 -0.02
        }
        device [
          PositionSensor {
            name "pos_rare"
          }
          RotationalMotor {
            name "motor_rear"
          }
        ]
        endPoint Solid {
          translation -1.3441096018019666e-07 -0.04000650958052026 -0.020051556435997354
          rotation 0.7026952567487779 0.11503792778476789 -0.7021293693570692 3.37081685509534
          children [
            DEF WHEEL Shape {
              appearance PBRAppearance {
                baseColor 0.333 0.333 0.333
                roughness 1
                metalness 0
              }
              geometry Cylinder {
                height 0.04
                radius 0.05
              }
            }
          ]
          name "wheel_rear"
          boundingObject USE WHEEL
          physics Physics {
            density -1
            mass 5
          }
        }
      }
      # --- Right wheel
      HingeJoint {
        jointParameters HingeJointParameters {
          position 226.02527724796843
          axis -1 0 0
          anchor 0.11 0.01 -0.02
        }
        device [
          PositionSensor {
            name "pos_right"
          }
          RotationalMotor {
            name "motor_right"
          }
        ]
        endPoint Solid {
          translation 0.11 0.01 -0.02
          rotation 0.08435980496045484 -0.9928634962409861 -0.08429413466641333 1.577157194145529
          children [
            USE WHEEL
          ]
          name "wheel_right"
          boundingObject USE WHEEL
          physics Physics {
            density -1
            mass 1
          }
        }
      }
      # --- Left wheel
      HingeJoint {
        jointParameters HingeJointParameters {
          position -69.68366539583312
          axis -1 0 0
          anchor -0.11 0.01 -0.02
        }
        device [
          PositionSensor {
            name "pos_left"
          }
          RotationalMotor {
            name "motor_left"
          }
        ]
        endPoint Solid {
          translation -0.11 0.01 -0.02
          rotation -0.27027449690492644 0.9241318308038782 0.2700593557310344 4.634364052336627
          children [
            USE WHEEL
          ]
          name "wheel_left"
          boundingObject USE WHEEL
          physics Physics {
            density -1
            mass 1
          }
        }
      }
      Pose {
        children [
          DEF BODY Shape {
            appearance PBRAppearance {
              baseColor 0.666 0.666 0.666
              roughness 1
              metalness 0
            }
            geometry Cylinder {
              height 0.1
              radius 0.1
            }
          }
        ]
      }
    ]
    boundingObject USE BODY
    physics Physics {}
  }
}
